### ISSUE TITLE
[BUG] CSP Adapter includes nodes from the local cluster

### ISSUE THREAD

**Issue by MbolotSuse** [rancher/rancher#38712]
**Rancher Server Setup**
- Rancher version: 2.6.7
- CSP Adapter version: 1.0.0 

**Information about the Cluster**
- Kubernetes version: 1.22
- Cluster Type (Local/Downstream): EKS, Local

**Describe the bug**
The csp adapter includes nodes from the local cluster in the total node counts. So if the local cluster has 2 nodes, a downstream cluster has 19 nodes, and I have 1 entitlement, the adapter considers the setup out of compliance (2 + 19 = 21 > 20). However, this setup should be in compliance, since nodes from the local cluster should not be counted toward totals.

**To Reproduce**
1. Deploy a 2 node EKS cluster and Rancher according to the adapter's [prerequisites](https://rancher.com/docs/rancher/v2.6/en/installation/cloud-marketplace/aws/prerequisites/).
2. Register one entitlement with AWS. 
3. Install the [adapter](https://rancher.com/docs/rancher/v2.6/en/installation/cloud-marketplace/aws/adapter-install/) according to the docs.
4. Deploy a downstream cluster with 19 nodes.

**Result**

A banner appears, and the csp adapter considers the setup out of compliance.

**Expected Result**

No banner appears, and the adapter does not consider the setup out of compliance.

**Comment by MbolotSuse** [comment]
## Validation Template

v1.0.1-rc1 has been published to dockerhub/ecr with these changes. Since the listing has not been updated, you will need to use the dockerhub version of the adapter when testing (i.e. add the `--set global.cattle.systemDefaultRegistry=docker.io` line when installing the adapter) as well as specifying a new tag to make it use this version rather than the 1.0.0 in the rancher/charts repo.

## Root Cause

The adapter was including node counts from any cluster in the node totals, rather than excluding the counts from the local cluster.

## What was fixed, or what change have occurred

- New logic has been added to ignore node counts from the local cluster
- If we can't verify which cluster a node count comes from, we also do not include it in the node total
- A related bug which prevented the adapter from functioning when it needed to check out 0 entitlements has also been fixed

## Areas or cases that should be tested

1. The test case outlined in the issue (i.e. where local + downstream would cause you to go over the current entitled limit) but just downstream doesn't cause you to go over the entitled limit).
2. Case where no downstream nodes are present (just local cluster). Ensure that there is no error message in the logs/banner, and the adapter's compliance message should state 0 entitlements required and 0 checked out.
3. Ensure that no earlier version of the adapter can be installed. In an account with access to 1.0.0 of the adapter, attempt to install version 1.0.1-rc1 or version 1.0.0-rc7 (override the image tag). This should fail with an `ErrImgPull` stating that the account does not have access to the product/image.
4. Ensure that basic adapter checkin/checkout logic works. For example -> buy 2 entitlements -> deploy 19 nodes -> ensure 1 entitlement is checked out (support config/aws console) -> increase to 21 nodes -> ensure 2 entitlements are checked out -> decrease back to 19 nodes -> ensure 1 entitlement is checked out -> delete the downstream cluster 

## What areas could experience regressions

Basic check-in/check-out functionality. See test case 4 for a validation to run to ensure that this continues to work.

## Are the repro steps accurate/minimal?

Yes - see the original issue.

**Comment by brudnak** [comment]
## Reproduction Environment

| Component                 | Version / Type                           |
|---------------------------|------------------------------------------|
| Rancher version           | 2.6.7                                    |
| Installation option       | Helm                                     |
| If Helm Chart k8s cluster | EKS v1.22.11-eks-18ef993                 |
| Cert Details              | Let's Encrypt                            |
| Helm version              | v2.16.8-rancher2                         |
| Downstream cluster type   | RKE1 Linode                              |
| Downstream K8s version    | v1.24.2-rancher1-1                       |
| Logged in user role       | Administrator & Standard User            |
| Browser type              | Google Chrome                            |
| Browser version           | 104.0.5112.101 (Official Build) (x86_64) |


**ADDITIONAL SETUP**

**EKS setup via `eksctl`**
```shell
eksctl create cluster \
  --name $NAME_VAR \
  --version 1.22 \
  --region $REGION_VAR \
  --nodegroup-name $NG_NAME_VAR \
  --nodes 3 \
  --nodes-min 1 \
  --nodes-max 4 \
  --managed
```

**Ingress installation**
```shell
helm repo add ingress-nginx https://kubernetes.github.io/ingress-nginx
helm repo update
helm upgrade --install \
  ingress-nginx ingress-nginx/ingress-nginx \
  --namespace ingress-nginx \
  --set controller.service.type=LoadBalancer \
  --version 4.0.18 \
  --create-namespace
```

**Rancher installation**
```shell
kubectl create namespace cattle-system

kubectl apply -f https://github.com/cert-manager/cert-manager/releases/download/v1.7.1/cert-manager.crds.yaml

helm install cert-manager jetstack/cert-manager \
  --namespace cert-manager \
  --create-namespace \
  --version v1.7.1
  
helm install rancher rancher-latest/rancher \
  --namespace cattle-system \
  --set hostname=$URL_VAR \
  --set bootstrapPassword=$PW_VAR \
  --set ingress.tls.source=letsEncrypt \
  --set letsEncrypt.email=$EMAIL_VAR \
  --set letsEncrypt.ingress.class=nginx \
  --set ingress.ingressClassName=nginx \
  --set extraEnv\[0\].name="CATTLE_PROMETHEUS_METRICS" --set-string extraEnv\[0\].value=true
```

*CSP Adapter Installation*
```shell
helm install rancher-csp-adapter rancher-dev/rancher-csp-adapter --namespace cattle-csp-adapter-system --set aws.enabled=true --set aws.roleName=$ROLE_VAR --set-string aws.accountNumber=$ACCOUNT_VAR --set global.cattle.systemDefaultRegistry=docker.io --create-namespace --set image.tag=v1.0.0 
```

## Reproduction steps

1. Follow instructions for setting up using BYOC (bring your own cluster) docs located at: https://rancher.com/docs/rancher/v2.6/en/installation/cloud-marketplace/aws
2. The big difference between reproduction and validation is that
   1. reproduction is using this line in the csp adapter: `--set image.tag=v1.0.0`
   2. validation is using this line in the csp adapter: `--set image.tag=v1.0.1`
3. Linked AWS account has 1 associated license for 20 nodes
4. Create a downstream RKE1 Linode cluster with 19 nodes (the local cluster has 3 nodes)
5. Check the UI for a compliance banner message
6. Download the support config and confirm Rancher is still in compliance
7. The local nodes are counting towards the total entitlements

## Additional Info

## RESULTS

#### ✅ Expected
For the local nodes not to count against total entitlements

#### ❌ Actual
Local nodes are counting towards total entitlements

<hr>

## Validation Environment

| Component                 | Version / Type                           |
|---------------------------|------------------------------------------|
| Rancher version           | 2.6.7                                    |
| Installation option       | Helm                                     |
| If Helm Chart k8s cluster | EKS v1.22.11-eks-18ef993                 |
| Cert Details              | Let's Encrypt                            |
| Helm version              | v2.16.8-rancher2                         |
| Downstream cluster type   | RKE1 Linode                              |
| Downstream K8s version    | v1.24.2-rancher1-1                       |
| Logged in user role       | Administrator & Standard User            |
| Browser type              | Google Chrome                            |
| Browser version           | 104.0.5112.101 (Official Build) (x86_64) |

**ADDITIONAL SETUP**

**EKS setup via `eksctl`**
```shell
eksctl create cluster \
  --name $NAME_VAR \
  --version 1.22 \
  --region $REGION_VAR \
  --nodegroup-name $NG_NAME_VAR \
  --nodes 3 \
  --nodes-min 1 \
  --nodes-max 4 \
  --managed
```

**Ingress installation**
```shell
helm repo add ingress-nginx https://kubernetes.github.io/ingress-nginx
helm repo update
helm upgrade --install \
  ingress-nginx ingress-nginx/ingress-nginx \
  --namespace ingress-nginx \
  --set controller.service.type=LoadBalancer \
  --version 4.0.18 \
  --create-namespace
```

**Rancher installation**
```shell
kubectl create namespace cattle-system

kubectl apply -f https://github.com/cert-manager/cert-manager/releases/download/v1.7.1/cert-manager.crds.yaml

helm install cert-manager jetstack/cert-manager \
  --namespace cert-manager \
  --create-namespace \
  --version v1.7.1
  
helm install rancher rancher-latest/rancher \
  --namespace cattle-system \
  --set hostname=$URL_VAR \
  --set bootstrapPassword=$PW_VAR \
  --set ingress.tls.source=letsEncrypt \
  --set letsEncrypt.email=$EMAIL_VAR \
  --set letsEncrypt.ingress.class=nginx \
  --set ingress.ingressClassName=nginx \
  --set extraEnv\[0\].name="CATTLE_PROMETHEUS_METRICS" --set-string extraEnv\[0\].value=true
```

*CSP Adapter Installation*
```shell
helm install rancher-csp-adapter rancher-dev/rancher-csp-adapter --namespace cattle-csp-adapter-system --set aws.enabled=true --set aws.roleName=$ROLE_VAR --set-string aws.accountNumber=$ACCOUNT_VAR --set global.cattle.systemDefaultRegistry=docker.io --create-namespace --set image.tag=v1.0.1-rc1 
```

## Reproduction steps

#### Test Case: Local Cluster Nodes Shouldn't Count Against Entitled Limit

1. Follow instructions for setting up using BYOC (bring your own cluster) docs located at: https://rancher.com/docs/rancher/v2.6/en/installation/cloud-marketplace/aws
2. Linked AWS account has 1 associated license for 20 nodes
3. Create a downstream RKE1 Linode cluster with 19 nodes (the local cluster has 3 nodes)
4. Check the UI for a compliance banner message
5. Download the support config and confirm Rancher is still in compliance
6. Check the CSP Adapter logs for any errors
7. Check the Rancher logs for any errors
   1. Starting from the **default** Rancher homepage `/dashboard/home`
   2. Click `hamburger menu` ==> `locl` ==> `Workload` ==> `Deployments` ==> filtered to `cattle-csp-adapter-system`
   3. Click `rancher-csp-adapter` from the list ==> click the `ellipsis (three dots)` next to the pod ==> click `View Logs`
   4. No errors in logs


## Additional Info

## RESULTS

#### ✅ Expected
Expected the local cluster nodes to not count against the total entitlements

#### ✅ Actual
The local cluster nodes were not counted against the total entitlements


## Additional Test Cases

<table>
    <tbody>
        <tr>
            <th>Test</th>
            <th>Steps</th>
            <th>Expected</th>
            <th>Actual</th>
            <th>Pass/Fail</th>
        </tr>
        <tr>
            <td>No downstream nodes are present</td>
            <td>
                <ol>
                    <li>Follow instructions for setting up using BYOC (bring your own cluster) docs located at: https://rancher.com/docs/rancher/v2.6/en/installation/cloud-marketplace/aws</li>
                    <li>Check the csp adapter logs for errors</li>
                    <li>Download the support config</li>
                </ol>
            </td>
            <td>The adapter's compliance message should state 0 entitlements required and 0 checked out</td>
            <td>Rancher server required 0 license(s) and was able to check out 0 license(s)</td>
            <td>✅ PASS</td>
        </tr>
        <tr>
            <td>Ensure that no earlier version of the adapter can be installed. In an account with access to 1.0.0 of the adapter, attempt to install version 1.0.1-rc1 or version 1.0.0-rc7 (override the image tag). This should fail with an ErrImgPull stating that the account does not have access to the product/image</td>
            <td>
                <ol>
                    <li>Follow instructions for setting up using BYOC (bring your own cluster) docs located at: https://rancher.com/docs/rancher/v2.6/en/installation/cloud-marketplace/aws</li>
                    <li>Remove the following set from the csp adapter helm installation command</li>
                    <li>"--set global.cattle.systemDefaultRegistry=docker.io"</li>
                    <li>Attempt to install the adapter with the set in addition to the default</li>
                    <li>"--set image.tag=v1.0.0"</li>
               </ol>
            </td>
            <td>Expected to not be able to use earlier versions from ECR</td>
            <td>You are able to use earlier versions of the adapter from ECR</td>
            <td>⚠️ Expected to not work per dev team</td>
        </tr>
        <tr>
            <td>Ensure that basic adapter checkin/checkout logic works: Exceed entitlement limit</td>
            <td>
                <ol>
                    <li>Follow instructions for setting up using BYOC (bring your own cluster) docs located at: https://rancher.com/docs/rancher/v2.6/en/installation/cloud-marketplace/aws</li>
                    <li>Linked AWS account has 1 associated license for 20 nodes</li>
                    <li>Create a downstream RKE1 Linode cluster with 19 nodes (the local cluster has 3 nodes)</li>
                    <li>Rancher is still in compliance because the local nodes are not counting</li>
                    <li>Create another downstream RKE1 Linode cluster with 5 nodes, this will take the cluster out of compliance</li>
                    <li>The cluster goes out of compliance</li>
                    <li>The UI banner appears stating Rancher is out of compliance</li>
                    <li>The support config shows: "Rancher server required 2 license(s) and was able to check out 1 license(s)"</li>
                </ol>
            </td>
            <td>Expected that going over the entitlement limit with downstream nodes that the Rancher server would warn of being out of compliance</td>
            <td>Going over the entitlement with downstream nodes caused the UI compliance warning to appear and the support config correctly reflected this as well</td>
            <td>✅ PASS</td>
        </tr>
        <tr>
            <td>Increase the SUSE Rancher Premium Support (non-EU and non-UK only) license amount in a Rancher that's out of compliance. Can use the test case above as a starting point</td>
            <td>
                <ol>
                    <li>Start with the Rancher server from the test case above that is out of compliance</li>
                    <li>Login to AWS, go to the AWS marketplace</li>
                    <li>Navigate to the SUSE Rancher Premium Support (non-EU and non-UK only)</li>
                    <li>Go to "Launch More Software" then "Subscribe"</li>
                    <li>In the AWS Marketplace increase the license amount from 1 ==> 2</li>
                </ol>
            </td>
            <td>The CSP adapter should detect the newly available additional license, check it out, and go back into compliance</td>
            <td>Rancher came back into compliance within 30 seconds of purchasing the additional licnese from the AWS marketplace</td>
            <td>✅ PASS</td>
        </tr>
        <tr>
            <td>Deleting a cluster on an out of compliance Rancher should bring it into compliance if it brings to node level to an acceptable level</td>
            <td>
                <ol>
                    <li>Follow instructions for setting up using BYOC (bring your own cluster) docs located at: https://rancher.com/docs/rancher/v2.6/en/installation/cloud-marketplace/aws</li>
                    <li>The AWS account being used has 2 licenses for a total of 40 nodes</li>
                    <li>Create a RKE1 Linode downstream with 19 nodes</li>
                    <li>Create a RKE1 Linode downstream with 5 nodes</li>
                    <li>Create a RKE1 Linode downstream with 18 nodes</li>
                    <li>The Rancher server goes out of compliance: Rancher server required 3 license(s) and was able to check out 2 license(s)</li>
                    <li>Delete the 1st Linode downstream cluster with 19 nodes</li>
                    <li>The cluster came back into compliance after deleting the 19 node downstream cluster</li>
                </ol>
            </td>
            <td>The Rancher server should go back into compliance when deleting a downstream cluster that brought it over the node limit</td>
            <td>The compliance banner in the UI went away, and the support config also showed the server to be in compliance</td>
            <td>✅ PASS</td>
        </tr>
        <tr>
            <td>Different downstream node types (non-local) count towards total entitlements (RKE2)</td>
            <td>
                <ol>
                    <li>Follow instructions for setting up using BYOC (bring your own cluster) docs located at: https://rancher.com/docs/rancher/v2.6/en/installation/cloud-marketplace/aws</li>
                    <li>Import a custom RKE2 cluster</li>
                    <li>Make sure the nodes count towards total entitlements</li>
                </ol>
            </td>
            <td>Expected imported RKE2 cluster to count towards total entitlements</td>
            <td>The imported RKE2 cluster nodes counted towards total entitlements</td>
            <td>✅ PASS</td>
        </tr>
    </tbody>
</table>

**Comment by brudnak** [comment]
Reopening so dev can comment on the second test in the table listed above:

>Ensure that no earlier version of the adapter can be installed. In an account with access to 1.0.0 of the adapter, attempt to install version 1.0.1-rc1 or version 1.0.0-rc7 (override the image tag). This should fail with an ErrImgPull stating that the account does not have access to the product/image 

The PASS/FAIL status listed in the table is for the 2nd test is: `Expected to not work per dev team`

Once dev comments on this test, QA is good to sign off/close the issue.

**Comment by MbolotSuse** [comment]
Given the complications with restricting the 1.0.0 release of the adapter, and the fact that this restriction will likely not lead to any code changes on the adapter side, we are going to work on restricting the 1.0.0 release independently of releasing version 1.0.1 of the adapter. I'm ok with considering the restriction as a "separate item to fix" rather than as a part of the 1.0.1 release.

**Comment by brudnak** [comment]
All other QA checks look good for the CSP Adapter. Since the business is aware about this note: https://github.com/rancher/rancher/issues/38712#issuecomment-1231829575

Then QA is okay to sign off on this issue.

**Comment by MbolotSuse** [comment]
Note: This issue was merged prior to the split between 2.6 and 2.7.

**Comment by MbolotSuse** [comment]
I believe that this would be good to release note as part of 2.6.9 so that users know not to use version 1.0.0 of the adapter. 

Something like the below would work:

Fixed an issue which caused version 1.0.0 of the csp adapter to incorrectly count nodes. Users should only use version 1.0.1 of the adapter or higher to ensure that node counts are accurate.
